### PR TITLE
save all rooms' costmatrixes into memory on early game stages

### PR DESCRIPTION
**Pull request by samogot**
do not sure if recalcing a needed at all

### CODE REVIEW

**Comment by democratic-collaboration** [comment]
[democratic collaboration](https://github.com/TooAngel/democratic-collaboration)
Approved reviews will speed up the merge, request changes will slow it down.

Please review the PR to help.

File: src/prototype_room_memory.js [hunk 1]
```diff
@@ -21,22 +21,29 @@
...
-  cache.rooms[this.name] = cache.rooms[this.name] || {
-    find: {},
-    costMatrix: {},
-    routing: {}
-  };
+  cache.rooms[this.name] = cache.rooms[this.name] || {};
+  cache.rooms[this.name].find = cache.rooms[this.name].find || {};
+  cache.rooms[this.name].costMatrix = cache.rooms[this.name].costMatrix || {};
+  cache.rooms[this.name].routing = cache.rooms[this.name].routing || {};
+};
```
Comment (TooAngel) [inline src/prototype_room_memory.js]: One reason for that kind of initialization was the codeclimate error.

File: src/prototype_room_memory.js [hunk 2]
```diff
@@ -21,22 +21,29 @@
  */
 Room.prototype.setMemoryCostMatrix = function(costMatrix) {
   this.checkCache();
-  if (this.controller && this.controller.my) {
+  if (this.controller && this.controller.my || Game.gcl.level < 15) {
```
Comment (TooAngel) [inline src/prototype_room_memory.js]: Can we extract this as a config variable.

File: src/prototype_room_memory.js [hunk 3]
```diff
@@ -52,6 +59,9 @@ Room.prototype.getMemoryCostMatrix = function() {
     if (!this.memory.costMatrix || !this.memory.costMatrix.base) {
       return;
     }
+    if (this.shouldRecalcCostmatrix()) {
```
Comment (TooAngel) [inline src/prototype_room_memory.js]: I'm not sure if the costMatrix needs to recalculated. The costMatrix should only be based on static room data, so it doesn't matter when you calculate it.

If we should introduce costmatrixes for owned and unowned rooms, I think we should trigger a recalculate when the room in claimed.

**Comment by samogot** [comment]
ready)

**Comment by democratic-collaboration** [comment]
A new review, yeah.

    Votes: 205/383
    Coefficient: 0.535248041775
    Merging in 3 days 19 hours
    Age 0 days 20 hours

**Pull request merged**